File: apps/agni-server/src/core/interactions/patrimony/removeSnapshotFromPatrimony.ts
import Repository from "@core/adapters/repository";
import { IUsecase } from "../interfaces";
import { ResourceNotFoundError } from "@core/errors/resournceNotFoundError";
import { PatrimonySnapshot } from "@core/domains/entities/patrimonySnapshot";


export class RemoveSnapshotFromPatrimonyUseCase implements IUsecase<string, void> {
    private snapshotRepo: Repository<PatrimonySnapshot>

    constructor(snapshotRepo: Repository<PatrimonySnapshot>) {
        this.snapshotRepo = snapshotRepo
    }

    async execute(id: string): Promise<void> {
        if (!await this.snapshotRepo.get(id))
            throw new ResourceNotFoundError("SNAPSHOT_PATRIMONY_NOT_FOUND")

        await this.snapshotRepo.delete(id)
    }

}

File: apps/agni_api/src/main/kotlin/dev/auguste/agni_api/core/usecases/patrimonies/snapshots/RemoveSnapshotFromPatrimony.kt
package dev.auguste.agni_api.core.usecases.patrimonies.snapshots

import dev.auguste.agni_api.core.adapters.repositories.IRepository
import dev.auguste.agni_api.core.entities.PatrimonySnapshot
import dev.auguste.agni_api.core.usecases.interfaces.IUseCase
import dev.auguste.agni_api.core.usecases.patrimonies.snapshots.dto.RemoveSnapshotFromPatrimonyInput
import java.util.UUID

class RemoveSnapshotFromPatrimony(private val snapshotRepo: IRepository<PatrimonySnapshot>): IUseCase<RemoveSnapshotFromPatrimonyInput, Unit> {

    override fun execAsync(input: RemoveSnapshotFromPatrimonyInput) {
        snapshotRepo.get(input.snapshotId) ?: throw Error("Snapshot not found")

        snapshotRepo.delete(input.snapshotId)
    }
}